make timestamp string format cassandra compatible

alezzqz opened the pull request:
when we convert timestamp into string it must look like: '2017-12-27T11:57:42.500Z'                
it concerns any conversion except JSON timestamp format                                                                
JSON string has space as time separator and must look like: '2017-12-27 11:57:42.500Z'
both formats always contain milliseconds and timezone specification

Fixes #14518
Fixes #7997

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2524/

tchaikov commented on types/types.cc:
```diff
@@ -89,18 +89,15 @@ time_point_to_string(const T& tp)
-    auto micros = millis * 1000;
-    return fmt::format("{}.{:06d}", to_string(tm), micros);
+    millis = std::round(millis * 1000) / 1000;
+    return fmt::format("{}.{:03d}Z", to_string(tm), millis);
```
this changes the output like: 
```diff
- "2011-02-03T04:05:00"
+ "2011-02-03T04:05:00.000Z"
```

so if we want to go in this direction. we might want to update the tests in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 accordingly

alezzqz replied:
sorry, 404

denesb replied:
@tchaikov scylla-dtest.git is private repository, it is not available to external contributors (not sure why).

tchaikov replied:
sorry, i just realized this.

nyh replied:
> @tchaikov scylla-dtest.git is private repository, it is not available to external contributors (not sure why).

It was like that from day one, I also don't know why. 
@alezzqz you don't need to worry too much, the better (in my opinion) functional test frameworks are all in scylla.git itself and all open-source.

tchaikov commented on types/types.cc:
```diff
@@ -89,18 +89,15 @@ time_point_to_string(const T& tp)
     }
-    auto micros = millis * 1000;
-    return fmt::format("{}.{:06d}", to_string(tm), micros);
+    millis = std::round(millis * 1000) / 1000;
```
what is this for?

alezzqz replied:
to round to 3 decimal places

tchaikov replied:
`millis` is a `long int`, so if we multiply it by 1000 if it does not overflow, the result would be `millis * 1000`. after dividing it by `1000`, the result should be `millis`. am i missing anything? or could you please raise an example?

alezzqz replied:
oops, my mistake, will fix
thanks

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

alezzqz commented:
changes:
- removed millis rounding cause it has no sense: the value can't have more than 3 digits
- refactored `time_point_to_string` function to simplify it

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2551/

denesb commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2551/

The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.

alezzqz commented:
> The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.

I'd fix these tests but have no access to them (

denesb commented:
> > The dtests mentioned by @tchaikov in https://github.com/scylladb/scylla-dtest/blob/next/json_test.py#L50 have failed. We will have to adjust those before this PR can be merged.
> 
> I'd fix these tests but have no access to them (

I will look into fixing these tests today. Will keep you posted.

denesb commented:
dtest PR is here: https://github.com/scylladb/scylla-dtest/pull/3316, once it goes in we can merge this PR (provided no new failures pop up).

denesb commented:
I will re-kick the CI once https://github.com/scylladb/scylla-dtest/pull/3316 goes in.

nyh commented:
@alezzqz if I understood correctly from @denesb, this patch also fixes the behavor of JSON. So does it also "Fixes #7997"? If it does, please add this phrase to the commit message, and also un-xfail all the tests we have for 7997 (if they, hopefully, start passing).

nyh reviewed (changes requested):
Looks good, but please check if it also Fixes #7997 - I believe it should (and if it doesn't, maybe this patch has a bug). If it makes the xfail-ing tests for #7997 pass, this patch should also remove those xfail tags. Thanks.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2585/

denesb commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2585/

Looks like my test fixes didn't propagate yet. @fruch how much time does it take for https://github.com/scylladb/scylla-dtest/pull/3316 to make its way into CI?

fruch commented:
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2585/
> 
> Looks like my test fixes didn't propagate yet. @fruch how much time does it take for https://github.com/scylladb/scylla-dtest/pull/3316 to make its way into CI?

It needs to pass next as any other scylla commit

alezzqz commented:
> @alezzqz if I understood correctly from @denesb, this patch also fixes the behavor of JSON. So does it also "Fixes #7997"? If it does, please add this phrase to the commit message, and also un-xfail all the tests we have for 7997 (if they, hopefully, start passing).

Yes, it seems JSON uses the same function to convert to string so it's behavior should be fixed as well.
I will mention it in commit description.

alezzqz commented on test/boost/expr_test.cc:
```diff
@@ -163,7 +163,7 @@ BOOST_AUTO_TEST_CASE(expr_printer_timestamp_test) {
         timestamp_type
     );
-    BOOST_REQUIRE_EQUAL(expr_print(timestamp_const), "'2011-03-02T03:05:00+0000'");
+    BOOST_REQUIRE_EQUAL(expr_print(timestamp_const), "'2011-03-02T03:05:00.000Z'");
```
In this test we compare timestamp value converted to string so it's just a way of CAST but you're right, we need to make commit message more correct. I will try to do.
Concerning other tests using +0000 - as I see all of them use this string values to parse and create timestamps and that format is correct for cassandra too. Do you think we'd better replace them to .000Z? It makes sense...

alezzqz replied:
Agree, will change tests now

nyh replied:
I'm not sure what you mean - I said that I thought the changes you already did to the tests were fine, you don't need to change them again. I was surprised why the **old** result (before your patch) was the way it was, but the new result seems fine.

alezzqz replied:
Okay, seems I didn't understand you properly. So I will change commit description only

nyh replied:
And please (unrelated to this thread of discussion) check that the xfailing tests for #7997 now pass, and remove their xfail mark.

alezzqz replied:
the problem is the test reproduces few issues (#7972, #7988, #7997, #8001) but we fixed only #7997 so removing xfail mark will bring to failing test. Maybe we should extract part for #7997 into separate test?

nyh replied:
You're right. So you can't remove the xfail, sorry. Just please remove the word "#7997" from the xfail line, and that's it (leave the "Reproduces:" line).
I would indeed be nice if you could extract the few lines that test 7997 to a separate test, e.g., put it in cql-pytest/test_json.py. Then you can be sure your patch fixes 7997 (and then please add "Fixes #7997" to the commit message). Thanks.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

alezzqz commented:
changes v3:
- changed commit and PR description to be more correct
- extracted reproducing #7997 code from unit test json_test.py:testToJsonFct to new test json_test.py::testTimestampToJson

denesb commented:
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

Kicked CI.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2610/

denesb commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2610/

The same json tests are still failing. Futhermore, a scylla sstable dump test is failing. I will look into it, as I will be off tomorrow, this will probably only happen on Monday.

nyh commented:
Please split the Fixes line into two separate Fixes lines, or github won't recognize these magic keywords. As the github documentation https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue explains: "_Use full syntax for each issue_".

nyh commented on test/cql-pytest/cassandra_tests/validation/entities/json_test.py:
```diff
@@ -849,11 +849,6 @@ def testToJsonFct(cql, test_keyspace):
-            # Reproduces #7997:
-            execute(cql, table, "INSERT INTO %s (k, timestampval) VALUES (?, ?)", 0, datetime(2014, 1, 1, 0, 0, 0))
-            assert_rows(execute(cql, table, "SELECT k, toJson(timestampval) FROM %s WHERE k = ?", 0), [0, "\"2014-01-01 00:00:00.000Z\""])
-
```
Please do not modify tests in the cassandra_tests/ directory - these are tests translated from the Cassandra unit tests, and I don't want to add new tests or delete pieces of tests in these files. Please leave this test unchanged, and the new test you added, testTimestampToJson, please don't put it in the cassandra_tests/ directory but rather in `cql-pytest/test_json.py`. The comment in this file explains its purpose, and among other things it includes : _small tests aiming to reproduce bugs discovered by bigger Cassandra tests._

nyh commented on test/cql-pytest/cassandra_tests/validation/entities/json_test.py:
```diff
@@ -929,6 +924,15 @@ def testToJsonFct(cql, test_keyspace):
             assert_rows(execute(cql, table, "SELECT k, toJson(durationval) FROM %s WHERE k = ?", 0), [0, "\"1y1mo2d10h5m\""])
 
+# Reproduces issue #7997
+def testTimestampToJson(cql, test_keyspace):
```
Please move this test to `cql-pytest/test_json.py` (for reasons I explained above).
Please note that that the coding convention in the general cql-pytest/ directory is different from cassandra_tests - in cassandra_tests I used a bunch of convenience functions to help me port the Java unit tests, but these aren't available and aren't needed in cql-pytest/test_json.py.

But beyond that, you should look at the content of the test and simplify it the bare minimum... This test does NOT need the create_type (I don't know why you needed it), and it shouldn't even create a new table just for this test - it should use the "table1" table that all the tests in the same file are using. This reduces the runtime of such a test from something like 0.1 seconds to something like 0.001 seconds, and makes me happy ;-)

nyh commented on test/cql-pytest/cassandra_tests/validation/entities/json_test.py:
```diff
@@ -929,6 +924,15 @@ def testToJsonFct(cql, test_keyspace):
 
+# Reproduces issue #7997
+def testTimestampToJson(cql, test_keyspace):
+    with create_type(cql, test_keyspace, "(a int, b uuid, c set<text>)") as type_name:
```
This type isn't used in this test and can be removed.

nyh reviewed (changes requested):
Looks good, but I still have some requests (sorry...) about the tests.
We also need to make the dtests pass before we can commit this PR - and @denesb said he'll do it.

I guess that like the C++ boost tests, the dtests "enshrined" a bunch of wrong (or at least, non-Cassandra-compatible) outputs, and nobody ever tested these tests against Cassandra so we didn't know. This is why I always insist that test/cql-pytest tests should also be run against Cassandra.

Also, as I asked (and explained) in previous comment, please split the Fixes line in the commit message into two separate lines. Thanks.

alezzqz commented on test/cql-pytest/cassandra_tests/validation/entities/json_test.py:
```diff
@@ -929,6 +924,15 @@ def testToJsonFct(cql, test_keyspace):
+                "timestampval timestamp)") as table:
+            execute(cql, table, "INSERT INTO %s (k, timestampval) VALUES (?, ?)", 0, datetime(2014, 1, 1, 0, 0, 0))
+            assert_rows(execute(cql, table, "SELECT k, toJson(timestampval) FROM %s WHERE k = ?", 0), [0, "\"2014-01-01T00:00:00.000Z\""])
+
```
Thanks. I run cassandra in docker container, is it possible to run this script using cassandra in container?

nyh replied:
The run-cassandra script is not much more than a convenience script that runs Cassandra for you in a temporary directory, and then runs "pytest" to run all the tests. If you know how to run Cassandra yourself (via docker or however you're used to running Cassandra), then just run it yourself, and then do something like
```
cd cqlpytest
pytest json_test.py::some_function_name
```

If your Cassandra isn't listening on localhost:9042 but rather on some other IP address or port (possibly even on another machine), just pass the "--host" and/or "--port" option to pytest, e.g.,
```
pytest --host 1.2.3.4 --port 4567 json_test.py::some_function_name
```

It's very easy and convenient.

alezzqz replied:
thanks, will try

alezzqz replied:
I've got an idea how to fix problem with JSON string format but not sure we should do it in this patch...

alezzqz replied:
I've got an idea how to fix JSON format, will add to this patch if you don't want to split these fixes into different patches

alezzqz replied:
I've added a function to support JSON timestamp string format. It's the simplest way I found.
I tried to add t`o_json_string() `method to `timestamp_date_base_class` and maybe it would look better a little but it required a lot of linked changes and I've chosen another way.
All tests are passed locally, including our json tests on Cassandra as we discussed above

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

alezzqz commented:
changes v4:
- returned back all changes to imported cassandra tests
- created new test for #7997 in test/cql-pytest/test_json.py
- changed "fixed" lines in commit descriptions
- rebased on latest master

tchaikov commented on types/types.cc:
```diff
@@ -78,26 +78,19 @@ time_point_to_string(const T& tp)
-    return fmt::format("{}.{:06d}", to_string(tm), micros);
+
+    auto year_digits = tm.tm_year >= -1900 ? 4 : 5;
+    return fmt::format("{:-0{}d}-{:02d}-{:02d}T{:02d}:{:02d}:{:02d}.{:03d}Z",
```
note to myself, we can drop the workaround in https://github.com/scylladb/scylla-dtest/blob/0324001e1d5275eb6d6a120018d51e7e0d4813e0/scylla_sstable_test.py#L176, once this change lands.

denesb replied:
In fact, the above test fails with this PR currently, so we will have to modify said test to work both with and without this PR, before this can go it. A CI is running currently and we should soon know the up-to-date list of dtest that sill require patching.

It is such a pain to do changes that require dtest changes too...

tchaikov replied:
yeah, i see. in the small time window. we need to at least appease both versions. lemme at least create a patch for the scylla_sstable_test.py . i really want to replace sstableloader in favor of `scylla dump-data`. as, IMO, a partial uuid-generation solution is not a solution ready for GA.

this should help us to partially address https://github.com/scylladb/scylla-tools-java/issues/333 -- one tiny step to ditch `tools/java`. lemme know if i can help.

tchaikov replied:
the dtest finished, see https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/

- [ ] [scylla_sstable_test.TestScyllaSstableDumpataAllDatatypes.test_sstabledump_all_datatypes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/testReport/junit/scylla_sstable_test/TestScyllaSstableDumpataAllDatatypes/FullDtest___full_split000___test_sstabledump_all_datatypes/) should be addressed by https://github.com/scylladb/scylla-dtest/pull/3328
- [ ] [update_cluster_layout_tests.TestUpdateClusterLayout.test_add_node_with_large_partition4](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/3163/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/FullDtest___full_split001___test_add_node_with_large_partition4/): i am still on it

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2636/

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request
**CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

alezzqz commented:
changes v5:
- added function to support cassandra compatible format for JSON timestamp strings
- modified commit and PR description

nyh commented on cql3/type_json.cc:
```diff
@@ -489,7 +489,7 @@ struct to_json_string_visitor {
     sstring operator()(const bytes_type_impl& t) { return quote_json_string("0x" + t.to_string(bv)); }
     sstring operator()(const boolean_type_impl& t) { return t.to_string(bv); }
-    sstring operator()(const timestamp_date_base_class& t) { return quote_json_string(t.to_string(bv)); }
+    sstring operator()(const timestamp_date_base_class& t) { return quote_json_string(timestamp_to_json_string(t, bv)); }
```
But neither of these changes are critical. Anyway, now I'm waiting for CI to pass (I started it now).

alezzqz replied:
> Looks good, just one thought: If the function is called timestamp_to_json_string, maybe it should return the full JSON string, i.e., with the double quotes? It can either call quote_json_string itself like you did, or just add the two quotes in the beginning and the end (there is no need to escape anything because you know you didn't use any special characters).

I thought about that but decided to keep code's common style: when we wrapping strings in quotes in visitor class.
I could do it just inside function if you think it's better.

> Another nitpick: Since this new function timestamp_to_json_string() is used only in this source file, maybe its definition can be in this file and not a different file (and it can be a static function)? Or it needs non-public APIs to be implemented?

Oh, that's would be much better, I agree. And I wanted to do that way. But this function uses things from different module so we should export something anyway. Better of two evils...
I could try implement dependent functionality just in place if you think it would be better

denesb commented:
> @scylladb/scylla-maint please review and trigger CI for this pull request **CI Build:** https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/parambuild/?PR_NUMBER=14726

Kicked CI.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/

denesb commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/

[Tests / Sanity Tests / json_test.TestsToJsonSelect.test_basic_data_types](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/testReport/junit/json_test/TestsToJsonSelect/Tests___Sanity_Tests___test_basic_data_types/)
    [Tests / Sanity Tests / json_test.TestsFromJsonInsert.test_basic_data_types](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2685/testReport/junit/json_test/TestsFromJsonInsert/Tests___Sanity_Tests___test_basic_data_types/)

I don't understand why these failed, I already prepared these tests. I will look into this.

tchaikov commented:
@denesb 

```diff
- "2011-02-03T04:05:00.000Z"
+ "2011-02-03 04:05:00.000Z"
```

tchaikov commented on types/types.cc:
```diff
@@ -78,26 +78,21 @@ time_point_to_string(const T& tp)
-    auto micros = millis * 1000;
-    return fmt::format("{}.{:06d}", to_string(tm), micros);
+
+    const auto time_separator = (use_time_separator) ? "T" : " ";
```
why quote `use_time_separator` with parenthesis ? also, a single char would suffice, no need to use a string literal here.

denesb replied:
Agree, no need to send a new version for this.

tchaikov replied:
yeah. please consider it a nitpick. i will create a follow-up change once this change lands.

denesb commented:
> @denesb
> 
> ```diff
> - "2011-02-03T04:05:00.000Z"
> + "2011-02-03 04:05:00.000Z"
> ```

Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.

tchaikov commented:
> > @denesb
> > ```diff
> > - "2011-02-03T04:05:00.000Z"
> > + "2011-02-03 04:05:00.000Z"
> > ```
> 
> Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.

this is exactly what @nyh complained at https://github.com/scylladb/scylladb/pull/14726#discussion_r1273773283 . Cassandra use a non-ISO 8601 format when serializing a datetime to JSON, but it format it with ISO 8601 when it prints it as non-JSON. see https://github.com/apache/cassandra/blob/d08fd1ebd83edc4730316c9630e60b8c0b0be89d/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L108-L126

denesb commented:
> > > @denesb
> > > ```diff
> > > - "2011-02-03T04:05:00.000Z"
> > > + "2011-02-03 04:05:00.000Z"
> > > ```
> > 
> > 
> > Thanks @tchaikov. @alezzqz, @nyh I lost track of the discussion, is this intended change? In that case we will need another round of update to this dtest.
> 
> this is exactly what @nyh complained at [#14726 (comment)](https://github.com/scylladb/scylladb/pull/14726#discussion_r1273773283) . Cassandra uses a non-ISO 8601 format when serializing a datetime to JSON, but it formats it with ISO 8601 when it prints it as non-JSON. see https://github.com/apache/cassandra/blob/d08fd1ebd83edc4730316c9630e60b8c0b0be89d/src/java/org/apache/cassandra/serializers/TimestampSerializer.java#L108-L126

:(
Ok, let me prepare a dtest PR.

denesb commented:
https://github.com/scylladb/scylla-dtest/pull/3341

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-staging/job/tchaikov/job/scylla-ci/14/

denesb commented:
All the prerequisites got in, CI passed, queuing.

nyh commented:
I want to comment that in retrospect, this fix is more important than we had realized and fixes a third bug which we didn't have listed - and had we known about it it might have caused us to backport this fix.

The third bug is that Scylla couldn't read the timestamps that it itself wrote. It wrote timestamps like `2009-12-17T00:26:29.805000` which then it couldn't read because 6 digits of precision aren't allowed for timestamps (Scylla reports "Milliseconds Exceed Expected Length"). We had a customer complaining about exactly this. This is more serious than the minor Cassandra compatibility issues of https://github.com/scylladb/scylladb/issues/14518
or https://github.com/scylladb/scylladb/issues/7997.

You can see in this patch all the "805000" being changed to "805" in the tests, but the importance of this was not realized, and it wasn't even mentioned in the commit message or any of the issues it referred to.

CC @ptrsmrn @alezzqz

nyh commented:
Opened https://github.com/scylladb/scylladb/issues/16575 to consider further backports
